File: src/helpers/editor/syntax/rust.ts
import store from "../../../store/index";
import { SyntaxHighlight, Token } from "../../../store/interfaces";

// Define Rust keywords
const RUST_KEYWORDS: string[] = [
  "as",
  "break",
  "const",
  "continue",
  "crate",
  "else",
  "enum",
  "extern",
  "false",
  "fn",
  "for",
  "if",
  "impl",
  "in",
  "let",
  "loop",
  "match",
  "mod",
  "move",
  "mut",
  "pub",
  "ref",
  "return",
  "self",
  "Self",
  "static",
  "struct",
  "super",
  "trait",
  "true",
  "type",
  "unsafe",
  "use",
  "where",
  "while",
  "async",
  "await",
  "dyn",
  "abstract",
  "final",
  "override",
];

function createToken(color: string, value: string): Token {
  return { color, value };
}

let syntaxHighlight: SyntaxHighlight | null = null;
let langTokenTypes: any | null = null;

function handleWhitespace(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  color: string
): number {
  const startPos = pos;
  while (pos < inputChars.length && /\s/.test(inputChars[pos])) {
    pos++;
  }
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleMacro(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  const startPos = pos;
  while (pos < inputChars.length && /[a-zA-Z0-9_]/.test(inputChars[pos])) {
    pos++;
  }
  if (inputChars[pos] === "!") {
    pos++; // Include the '!' in the macro name
    const value = inputChars.slice(startPos, pos).join("");
    tokens.push(createToken(color, value));
  }
  return pos;
}

function handleLifetime(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  const startPos = pos;
  pos++; // Move past the apostrophe
  while (pos < inputChars.length && /[a-zA-Z0-9]/.test(inputChars[pos])) {
    pos++;
  }
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleStringLiteral(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  color: string
): number {
  const startChar = inputChars[pos];
  const startPos = pos;
  pos++; // Move past the opening quote

  if (startChar === "r") {
    // Handle raw strings in Rust
    let hashCount = 0;
    while (inputChars[pos] === "#") {
      hashCount++;
      pos++;
    }

    if (inputChars[pos] === `"`) {
      pos++; // Move past opening quote
      const closingSequence = `"` + "#".repeat(hashCount);

      // Find the closing raw string sequence
      while (pos + hashCount < inputChars.length) {
        if (inputChars.slice(pos, pos + closingSequence.length).join("") === closingSequence) {
          pos += closingSequence.length;
          break;
        }
        pos++;
      }
    }
  } else if (startChar === "'") {
    // Handle single-character literals
    while (pos < inputChars.length && inputChars[pos] !== "'") {
      if (inputChars[pos] === "\\" && pos + 1 < inputChars.length) pos += 2;
      else pos++;
    }
    pos++;
  } else {
    // Handle double-quoted strings
    while (pos < inputChars.length && inputChars[pos] !== '"') {
      if (inputChars[pos] === "\\" && pos + 1 < inputChars.length) pos += 2;
      else pos++;
    }
    pos++;
  }

  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleAttribute(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  color: string
): number {
  const startPos = pos;
  pos++; // Move past `#`

  if (inputChars[pos] === "[") {
    pos++;
    while (pos < inputChars.length && inputChars[pos] !== "]") pos++;
    pos++;
  }

  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleComment(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  const startPos = pos;
  while (pos < inputChars.length && inputChars[pos] !== "\n") pos++;
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleNumber(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  const startPos = pos;
  while (pos < inputChars.length && /\d/.test(inputChars[pos])) pos++;
  const value = inputChars.slice(startPos, pos).join("");
  tokens.push(createToken(color, value));
  return pos;
}

function handleOperator(tokens: Token[], inputChars: string[], pos: number, color: string): number {
  let value = inputChars[pos];
  pos++;

  if (pos < inputChars.length) {
    const nextChar = inputChars[pos];
    if (value === "=" && (nextChar === "=" || nextChar === ">")) {
      value += nextChar;
      pos++;
    } else if (value === "!" && nextChar === "=") {
      value += nextChar;
      pos++;
    }
  }

  tokens.push(createToken(color, value));
  return pos;
}

function handleIdentifier(
  tokens: Token[],
  inputChars: string[],
  pos: number,
  functionNameColor: string,
  keywordColor: string,
  identifierColor: string
): number {
  const startPos = pos;
  while (pos < inputChars.length && /[a-zA-Z0-9_]/.test(inputChars[pos])) pos++;
  const value = inputChars.slice(startPos, pos).join("");
  const nextChar = inputChars[pos] || " ";

  if (nextChar === "(") {
    tokens.push(createToken(functionNameColor, value));
  } else if (RUST_KEYWORDS.includes(value)) {
    tokens.push(createToken(keywordColor, value));
  } else {
    tokens.push(createToken(identifierColor, value));
  }

  return pos;
}

function handlePunctuation(
  tokens: Token[],
  currentChar: string,
  pos: number,
  color: string
): number {
  tokens.push(createToken(color, currentChar));
  return pos + 1;
}

function handleUnknown(tokens: Token[], currentChar: string, pos: number, color: string): number {
  tokens.push(createToken(color, currentChar));
  return pos + 1;
}

export function tokenizeRustCode(input: string): Token[] | null {
  const tokens: Token[] = [];

  if (!syntaxHighlight || !langTokenTypes) {
    syntaxHighlight = store.state.settings.syntaxHighlight;
    if (!syntaxHighlight) return null;

    langTokenTypes = syntaxHighlight.languages.find((lang) => lang.language === "rust");
    if (!langTokenTypes) return null;
  }

  let pos = 0;
  const inputChars = Array.from(input);

  while (pos < inputChars.length) {
    const currentChar = inputChars[pos];

    if (/\s/.test(currentChar)) {
      pos = handleWhitespace(tokens, inputChars, pos, langTokenTypes.whitespaceColor);
    } else if (currentChar === "#") {
      pos = handleAttribute(tokens, inputChars, pos, langTokenTypes.attributeColor);
    } else if (
      currentChar === "/" &&
      (inputChars[pos + 1] === "/" || inputChars[pos + 1] === "*")
    ) {
      pos = handleComment(tokens, inputChars, pos, langTokenTypes.commentColor);
    } else if (currentChar === '"' || currentChar === "'") {
      pos = handleStringLiteral(tokens, inputChars, pos, langTokenTypes.stringLiteralColor);
    } else if (currentChar === "'" && /[a-zA-Z0-9]/.test(inputChars[pos + 1])) {
      pos = handleLifetime(tokens, inputChars, pos, langTokenTypes.lifetimeColor);
    } else if (/\d/.test(currentChar)) {
      pos = handleNumber(tokens, inputChars, pos, langTokenTypes.numberColor);
    } else if ("+-*/%=><!".includes(currentChar)) {
      pos = handleOperator(tokens, inputChars, pos, langTokenTypes.operatorColor);
    } else if (/[a-zA-Z_]/.test(currentChar)) {
      pos = handleIdentifier(
        tokens,
        inputChars,
        pos,
        langTokenTypes.functionNameColor,
        langTokenTypes.keywordColor,
        langTokenTypes.identifierColor
      );
    } else if (",.:;()[]{}<>".includes(currentChar)) {
      pos = handlePunctuation(tokens, currentChar, pos, langTokenTypes.punctuationColor);
    } else {
      pos = handleUnknown(tokens, currentChar, pos, langTokenTypes.unknownColor);
    }
  }
  return tokens;
}

File: src-tauri/src/syntax_highlight/languages/rust.rs
use crate::settings::Settings;
use crate::syntax_highlight::{Token, TokenType, create_token};

const RUST_KEYWORDS: [&str; 41] = [
    "as", "break", "const", "continue", "crate", "else", "enum", "extern",
    "false", "fn", "for", "if", "impl", "in", "let", "loop", "match",
    "mod", "move", "mut", "pub", "ref", "return", "self", "Self", "static",
    "struct", "super", "trait", "true", "type", "unsafe", "use", "where",
    "while", "async", "await", "dyn", "abstract", "final", "override"
];

fn handle_macro(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && (input_chars[*pos].is_alphanumeric() || input_chars[*pos] == '_') {
        *pos += 1;
    }
    if *pos < input_chars.len() && input_chars[*pos] == '!' {
        *pos += 1; // Include the '!' in the macro name
        let value: String = input_chars[start_pos..*pos].iter().collect();
        tokens.push(create_token(settings_config, "rust", TokenType::FunctionName, &value));
    }
}

fn handle_lifetime(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    *pos += 1; // Move past the apostrophe
    while *pos < input_chars.len() && input_chars[*pos].is_alphanumeric() {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "rust", TokenType::Identifier, &value));
}

fn handle_string_literal(
    tokens: &mut Vec<Token>, 
    settings: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_char = input_chars[*pos];
    let start_pos = *pos;
    *pos += 1; // Move past the opening quote

    if start_char == 'r' {
        // Handle raw strings in Rust
        let mut hash_count = 0;
        while *pos < input_chars.len() && input_chars[*pos] == '#' {
            hash_count += 1;
            *pos += 1;
        }

        // Check if it's a raw string with double quotes after the hashes
        if *pos < input_chars.len() && input_chars[*pos] == '"' {
            *pos += 1; // Move past the opening quote
            let closing_sequence: Vec<char> = std::iter::repeat('#').take(hash_count).collect();

            // Process until we find the closing raw string sequence
            while *pos + hash_count < input_chars.len() {
                if input_chars[*pos] == '"' && &input_chars[*pos + 1..*pos + 1 + hash_count] == closing_sequence.as_slice() {
                    *pos += 1 + hash_count; // Move past the closing quote and hashes
                    break;
                }
                *pos += 1;
            }
        }
    } else if start_char == '\'' {
        // Handle single-quoted characters
        while *pos < input_chars.len() && input_chars[*pos] != '\'' {
            // Allow escaped single-quote within single character
            if input_chars[*pos] == '\\' && *pos + 1 < input_chars.len() {
                *pos += 2;
                continue;
            }
            *pos += 1;
        }
        *pos += 1; // Move past the closing single quote
    } else {
        // Handle double-quoted strings
        while *pos < input_chars.len() && input_chars[*pos] != '"' {
            // Allow for escaped quotes within double-quoted strings
            if input_chars[*pos] == '\\' && *pos + 1 < input_chars.len() {
                *pos += 2;
                continue;
            }
            *pos += 1;
        }
        *pos += 1; // Move past the closing double quote
    }

    // Collect the entire string value including quotes
    let value: String = input_chars[start_pos..*pos.min(&mut input_chars.len())].iter().collect();
    tokens.push(create_token(settings, "rust", TokenType::StringLiteral, &value));
}


fn handle_attribute(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    *pos += 1; // Move past the opening `#`

    if *pos < input_chars.len() && input_chars[*pos] == '[' {
        *pos += 1;
        while *pos < input_chars.len() && input_chars[*pos] != ']' {
            *pos += 1;
        }
        *pos += 1; // Move past the closing `]`
    }

    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "rust", TokenType::FunctionName, &value));
}

fn handle_whitespace(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && input_chars[*pos].is_whitespace() {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "rust", TokenType::Whitespace, &value));
}

fn handle_comment(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && input_chars[*pos] != '\n' {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "rust", TokenType::Comment, &value));
}

fn handle_number(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && input_chars[*pos].is_digit(10) {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    tokens.push(create_token(settings_config, "rust", TokenType::Number, &value));
}

fn handle_operator(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    current_char: char, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let mut value = current_char.to_string();
    *pos += 1;

    // Handle multi-character operators like ==, ===, =>, etc.
    if let Some(next_char) = input_chars.get(*pos) {
        if current_char == '=' && (*next_char == '=' || *next_char == '>') {
            value.push(*next_char);
            *pos += 1;

            // Handle === case
            if *next_char == '=' && *pos < input_chars.len() && input_chars[*pos] == '=' {
                value.push('=');
                *pos += 1;
            }
        } else if current_char == '!' && *next_char == '=' {
            value.push(*next_char);
            *pos += 1;
        }
    }

    tokens.push(create_token(settings_config, "rust", TokenType::Operator, &value));
}

fn handle_identifier(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    input_chars: &[char], 
    pos: &mut usize
) {
    let start_pos = *pos;
    while *pos < input_chars.len() && (input_chars[*pos].is_alphanumeric() || input_chars[*pos] == '_') {
        *pos += 1;
    }
    let value: String = input_chars[start_pos..*pos].iter().collect();
    let next_char = input_chars.get(*pos).unwrap_or(&' ');

    // Identify function calls and class names in JavaScript
    if *next_char == '(' {
        tokens.push(create_token(settings_config, "rust", TokenType::FunctionName, &value));
    } else {
        let token_type;
        if RUST_KEYWORDS.contains(&value.as_str()) {
            token_type = TokenType::Keyword;
        } else if value.chars().next().unwrap().is_uppercase() {
            token_type = TokenType::ClassName;
        } else {
            token_type = TokenType::Identifier;
        }
        tokens.push(create_token(settings_config, "rust", token_type, &value));
    }
}

fn handle_punctuation(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    current_char: char, 
    pos: &mut usize
) {
    let value = current_char.to_string();
    tokens.push(create_token(settings_config, "rust", TokenType::Punctuation, &value));
    *pos += 1;
}

fn handle_unknown(
    tokens: &mut Vec<Token>, 
    settings_config: &Settings, 
    current_char: char, 
    pos: &mut usize
) {
    let value = current_char.to_string();
    tokens.push(create_token(settings_config, "rust", TokenType::Unknown, &value));
    *pos += 1;
}

pub fn tokenize_rust_code(settings_config: &Settings, input: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    let mut pos = 0;
    let input_chars: Vec<char> = input.chars().collect();

    while pos < input_chars.len() {
        let current_char = input_chars[pos];

        if current_char.is_whitespace() {
            handle_whitespace(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '#' {
            handle_attribute(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '/' && (input_chars.get(pos + 1) == Some(&'/') || input_chars.get(pos + 1) == Some(&'*')) {
            handle_comment(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '"' || current_char == '\'' {
            handle_string_literal(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char == '\'' && input_chars.get(pos + 1).map(|c| c.is_alphanumeric()).unwrap_or(false) {
            handle_lifetime(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if current_char.is_digit(10) {
            handle_number(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
        } 
        else if "+-*/=%><!".contains(current_char) {
            handle_operator(&mut tokens, settings_config, current_char, input_chars.as_slice(), &mut pos);
        } 
        else if current_char.is_alphabetic() || current_char == '_' {
            if input_chars.get(pos + 1) == Some(&'!') {
                handle_macro(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
            } else {
                handle_identifier(&mut tokens, settings_config, input_chars.as_slice(), &mut pos);
            }
        } 
        else if ",.:;()[]{}<>".contains(current_char) {
            handle_punctuation(&mut tokens, settings_config, current_char, &mut pos);
        } 
        else {
            handle_unknown(&mut tokens, settings_config, current_char, &mut pos);
        }
    }

    tokens
}
